src/Catalog/Helpers/GalleryDatabaseQueryService.cs

```diff
@@ -0,0 +1,166 @@
+            foreach (var keyDate in result.Keys)
+            {
+                if (result.TryGetValue(keyDate, out IList<FeedPackageDetails> packagesForKeyDate))
+                {
+                    if (packagesCount > 0 && packagesCount + packagesForKeyDate.Count > Constants.MaxPageSize)
+                    {
+                        break;
+                    }
+
+                    packagesCount += packagesForKeyDate.Count;
+                    filteredResult.Add(keyDate, packagesForKeyDate);
+                }
+            }
+
+            return filteredResult;
+        }
+
+        /// <summary>
+        /// Builds the SQL query string for db2catalog.
+        /// </summary>
+        /// <param name="cursor">The <see cref="DB2CatalogCursor"/> to be used.</param>
+        /// <returns>The SQL query string for the db2catalog job, build from the the provided <see cref="DB2CatalogCursor"/>.</returns>
+        internal static string BuildDB2CatalogSqlQuery(DB2CatalogCursor cursor)
+        {
+            return $@"SELECT TOP {cursor.Top} WITH TIES
+                        PR.[Id],
+                        P.[NormalizedVersion],
+                        P.[Created],
+                        P.[LastEdited],
+                        P.[Published],
+                        P.[Listed],
+                        P.[HideLicenseReport],
+                        P.[LicenseNames],
+                        P.[LicenseReportUrl]
+                    FROM [dbo].[Packages] AS P
+                    INNER JOIN [dbo].[PackageRegistrations] AS PR ON P.[PackageRegistrationKey] = PR.[Key]
+                    WHERE P.[PackageStatusKey] = {PackageStatus.Available} 
+                        AND P.[{cursor.ColumnName}] > @{CursorParameterName}
+                    ORDER BY P.[{cursor.ColumnName}]";
+        }
+
+        /// <summary>
+        /// Asynchronously gets a <see cref="IReadOnlyCollection{FeedPackageDetails}"/> from the gallery database.
+        /// </summary>
+        /// <param name="cursor">Defines the cursor to be used.</param>
+        /// <returns>A task that represents the asynchronous operation.
+        /// The task result (<see cref="Task{TResult}.Result" />) returns an
+        /// <see cref="IReadOnlyCollection{FeedPackageDetails}" />.</returns>
+        private async Task<IReadOnlyCollection<FeedPackageDetails>> GetPackages(DB2CatalogCursor cursor)
+        {
+            using (var sqlConnection = await _connectionFactory.OpenAsync())
+            {
+                return await GetPackageDetailsAsync(sqlConnection, cursor);
+            }
+        }
+
+        private async Task<IReadOnlyCollection<FeedPackageDetails>> GetPackageDetailsAsync(
+            SqlConnection sqlConnection,
+            DB2CatalogCursor cursor)
+        {
+            var packages = new List<FeedPackageDetails>();
+            var packageQuery = BuildDB2CatalogSqlQuery(cursor);
+
+            var packagesCommand = new SqlCommand(packageQuery, sqlConnection)
```

Comment:
Dispose

Reply:
Fixed in https://github.com/NuGet/NuGet.Services.Metadata/pull/524/commits/6a4b503ac01e97f319c50333e653fafd0fed1630